src/java/org/apache/cassandra/config/DatabaseDescriptor.java

```diff
@@ -1592,7 +1592,7 @@ public static Collection<String> tokensFromString(String tokenString)
 
     public static int getNumTokens()
     {
-        return conf.num_tokens;
+        return conf.num_tokens == null ? -1 : conf.num_tokens;
```

Comment:
This changes the semantics of the function.

Do we know that all the clients now handle `-1` and know how to deal with it?
Like - in what way returning `-1` makes the clients work correctly, if they never supported lack of value (NPE)?

Why `-1` and not `-12345`? If the value is irrelevant, then why are we calling the function at all?

Reply:
All the usages of this call check the return value and (in most cases) will throw `ConfigurationException` errors on the `-1` value. The problem here is that the compactor is a special case and effectively uses an offline configuration that doesn't configure tokens or use them. All the services that use online configuration are unaffected by this change because they all set tokens. 

This change effectively duplicates the implementation that is in DSE except that DSE has `num_tokens` defined as a `int`  and defaults it to `-1` in `Config`.

Reply:
If it's because of special case of compactor, couldn't we just:
`DatabaseDescriptor.getRawConfig().num_tokens = 1` 
before we create the offline configuration in `CompactorCli`?

> This change effectively duplicates the implementation that is in DSE except that DSE has num_tokens defined as a int and defaults it to -1 in Config.

This is not true. DSE 5.1 defaults to `1` in `Config`, DSE 6.8 defaults to `-1` in `Config`, but never returns `-1` from this function. Cassandra doesn't set the default, but has:
http://github.com/datastax/cassandra/blob/d211a70bf9fcf14d57851064cce46fa7bdc6b1d3/src/java/org/apache/cassandra/config/DatabaseDescriptor.java#L1212-L1238
which will set `num_tokens` to `1` typically

> All the usages of this call check the return value and (in most cases) will throw ConfigurationException errors on the -1 value.
What about the future clients, that might be added upstream? Are we going to monitor them and adjust on our fork if needed?

> The problem here is that the compactor is a special case and effectively uses an offline configuration that doesn't configure tokens or use them
I understand this as - the compactor doesn't need them, but unfortunately uses them (in `getDiskBoundaries`)

I feel uneasy about this change because it will make it easy for us to diverge semantically from upstream (and I don't see a drawback of fixing that in CNDB)

Reply:
Actually, perhaps we should set `num_tokens` here:

http://github.com/riptano/cndb/blob/e5ebf95d4ced516b25ff16dd60958045bca6a067/cndb-common/src/main/java/com/datastax/cndb/config/ConfigurationControl.java#L407-L408

Reply:
Yep, I think you are right here. It makes more sense to make sure that the offline configuration is correctly configured. I'll make the change there and drop this PR. I'll check our specific usage but I suspect the correct approach is to set the num_tokens to 1 there.